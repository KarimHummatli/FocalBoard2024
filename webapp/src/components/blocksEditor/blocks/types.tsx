import {ContentBlockTypes} from '../../../blocks/block'

export type BlockInputProps = {
    onChange: (value: string) => void
    value: string
    onCancel: () => void
    onSave: (val: string) => void
}

export type ContentType = {
    name: ContentBlockTypes
    displayName: string
    slashCommand: string
    prefix: string
    Input: React.FunctionComponent<BlockInputProps>
    render: (value: string) => React.ReactNode
    runSlashCommand: (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]) => void
    nextType?: string
}
