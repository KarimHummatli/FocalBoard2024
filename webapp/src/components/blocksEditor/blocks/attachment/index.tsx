import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './attachment.scss'

const Attachment: ContentType = {
    name: 'attachment',
    displayName: 'Attachment',
    slashCommand: '/attachment',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps) => <div className='AttachmentView'>ðŸ“Ž {props.value}</div>,
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Attachment'
                type='file'
                onChange={(e) => {
                    const files = e.currentTarget?.files
                    if (files) {
                        for (let i = 0; i < files.length; i++) {
                            const file = files.item(i)
                            if (file) {
                                props.onSave(file.name as string)
                            }
                        }
                    }
                }}
            />
        )
    }
}

Attachment.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Attachment)
    changeValue('')
}

export default Attachment
