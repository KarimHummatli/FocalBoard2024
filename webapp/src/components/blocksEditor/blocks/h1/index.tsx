import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h1.scss'

const H1: ContentType = {
    name: 'h1',
    displayName: 'Title',
    slashCommand: '/title',
    prefix: '# ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('# '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H1'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H1.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H1)
    changeValue(args.join(' '))
}

export default H1
