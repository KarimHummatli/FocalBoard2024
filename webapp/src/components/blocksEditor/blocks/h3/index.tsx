import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h3.scss'

const H3: ContentType = {
    name: 'h3',
    displayName: 'Sub Sub title',
    slashCommand: '/subsubtitle',
    prefix: '### ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('### '+value, {renderer, breaks: true})
        return <div
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H3'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === "Backspace") {
                        props.onCancel()
                    }
                    if (e.key === "Enter") {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    }
}

H3.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H3)
    changeValue(args.join(' '))
}

export default H3
