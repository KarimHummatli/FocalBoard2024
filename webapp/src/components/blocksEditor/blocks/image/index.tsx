import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './image.scss'

type FileInfo = {
    file: string|ArrayBuffer
    width?: number
    align?: 'left'|'center'|'right'
}

const Image: ContentType<FileInfo> = {
    name: 'image',
    displayName: 'Image',
    slashCommand: '/image',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        if (props.value.file) {
            return <img className='ImageView' src={props.value.file as string}/>
        }
        return null
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <div>
                {props.value.file && (typeof props.value.file === 'string') && (
                    <img
                        className='ImageView' src={props.value.file}
                        onClick={() => ref.current?.click()}
                    />
                )}
                <input
                    ref={ref}
                    className='Image'
                    type='file'
                    accept='image/*'
                    onChange={(e) => {
                        const reader = new FileReader();
                        const file = (e.currentTarget?.files || [])[0]
                        reader.readAsArrayBuffer(file);
                        reader.onload = function () {
                            props.onSave({file: reader.result as ArrayBuffer})
                        }
                    }}
                />
            </div>
        )
    }
}

Image.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Image)
    changeValue({file: ''})
}

export default Image
