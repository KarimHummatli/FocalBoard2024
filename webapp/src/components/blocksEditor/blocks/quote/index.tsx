import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './quote.scss'

const Quote: ContentType = {
    name: 'quote',
    displayName: 'Quote',
    slashCommand: '/quote',
    prefix: '> ',
    render: (value: string) => {
        const renderer = new marked.Renderer()
        const html = marked('> '+value, {renderer, breaks: true})
        return <div
            className='Quote'
            dangerouslySetInnerHTML={{__html: html.trim()}}
        />
    },
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <blockquote className='Quote'>
                <input
                    ref={ref}
                    onChange={(e) => props.onChange(e.currentTarget.value)}
                    onKeyDown={(e) => {
                        if (props.value === '' && e.key === "Backspace") {
                            props.onCancel()
                        }
                        if (e.key === "Enter") {
                            props.onSave(props.value)
                        }
                    }}
                    onBlur={() => props.onSave(props.value)}
                    value={props.value}
                />
            </blockquote>
        )
    }
}

Quote.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(Quote)
    changeValue(args.join(' '))
}

export default Quote
