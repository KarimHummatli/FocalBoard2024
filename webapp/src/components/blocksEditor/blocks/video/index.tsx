import React, {useRef, useEffect} from 'react'
import {BlockInputProps, ContentType} from '../types'

import './video.scss'

const Video: ContentType = {
    name: 'video',
    displayName: 'Video',
    slashCommand: '/video',
    prefix: '',
    render: (value: string) => (
        <video width="320" height="240" controls className='VideoView'>
            <source src={value} type='video/mp4'/>
        </video>
    ),
    runSlashCommand: (): void => {},
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Video'
                type='file'
                accept='video/*'
                onChange={(e) => {
                    const reader = new FileReader();
                    const file = (e.currentTarget?.files || [])[0]
                    reader.readAsDataURL(file);
                    reader.onload = function () {
                        props.onSave(reader.result as string)
                    }
                }}
            />
        )
    }
}

Video.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void): void => {
    changeType(Video)
    changeValue('')
}

export default Video
