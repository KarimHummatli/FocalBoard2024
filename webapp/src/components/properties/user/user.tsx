// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useContext} from 'react'
import Select from 'react-select'

import {IUser, WorkspaceUsersContext, WorkspaceUsersContextData} from '../../../user'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
    tooltip?: string
}

const UserProperty = (props: Props): JSX.Element => {
    const workspaceUsers = useContext<WorkspaceUsersContextData>(WorkspaceUsersContext)
    const tooltipClassName = props.tooltip ? 'octo-tooltip tooltip-top' : ''

    if (props.readonly) {
        const userClassName = `UserProperty octo-propertyvalue ${tooltipClassName}`
        return (
            <div
                className={userClassName}
                data-tooltip={props.tooltip}
            >
                {workspaceUsers.usersById.get(props.value)?.username || props.value}</div>)
    }

    return (
        <Select
            options={workspaceUsers.users}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={'UserProperty'}
            styles={getSelectBaseStyle()}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={workspaceUsers.usersById.get(props.value) || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
