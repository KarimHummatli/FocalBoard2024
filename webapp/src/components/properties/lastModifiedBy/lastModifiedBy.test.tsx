// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {render} from '@testing-library/react'
import configureStore from 'redux-mock-store'

import {Card} from '../../../blocks/card'
import {IUser} from '../../../user'

import {Board} from '../../../blocks/board'

import {CommentBlock} from '../../../blocks/commentBlock'

import LastModifiedBy from './lastModifiedBy'

describe('components/properties/lastModifiedBy', () => {
    test('should match snapshot', () => {
        const card = new Card()
        card.id = 'card-id-1'
        card.modifiedBy = 'user-id-1'

        const board = new Board()
        const comment = new CommentBlock()
        comment.modifiedBy = 'user-id-1'
        comment.parentId = 'card-id-1'

        const mockStore = configureStore([])
        const store = mockStore({
            currentWorkspaceUsers: {
                byId: {
                    'user-id-1': {username: 'username_1'} as IUser,
                },
            },
        })

        const component = (
            <ReduxProvider store={store}>
                <LastModifiedBy
                    card={card}
                    board={board}
                    contents={[]}
                    comments={[comment]}
                />
            </ReduxProvider>
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })
})
