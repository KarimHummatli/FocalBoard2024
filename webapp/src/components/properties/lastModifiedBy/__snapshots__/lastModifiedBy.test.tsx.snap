// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/properties/lastModifiedBy should match snapshot 1`] = `
<div>
  <div>
    <div
      class="UserProperty octo-propertyvalue readonly"
    />
  </div>
</div>
`;
